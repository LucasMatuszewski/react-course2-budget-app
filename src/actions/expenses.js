import uuid from 'uuid';

// we make here NAMED EXPORTS (not default)

/***********
 * ACTIONS (action generators - functions to generate an Action)
 */

// ADD_EXPENSE ( property = default value )
export const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
    } = {}
) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(), // we use uuid - NPM library to generate RFC4122 UUIDs (we have to install and import it)
        description,
        note,
        amount,
        createdAt
    }
});

// REMOVE_EXPENSE
// const removeExpense = (id) => ({ // return object: {id: "UUID"}
export const removeExpense = ({ id } = {}) => ({ // return only UUID
    type: 'REMOVE_EXPENSE',
    id
    // expense: { id } // we don't need "expense: { }" here. We only want to pass id to delete it
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});