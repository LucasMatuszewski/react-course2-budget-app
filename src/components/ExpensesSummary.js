import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

// MY VERSION:
// 1. Summary Component inside ExpenseList with Filtered Expenses as props
// 2. I don't selectExpenses here to filter expenses (why do it twice if we do this in ExpenseList?)
export const ExpensesSummary = (props) => (
    <div>
        {
            props.expenseCount > 0 && (
                <p>
                    Total: {props.expensesTotal}, 
                    from {props.expenseCount} expense{ props.expenseCount !== 1 && 's' }
                </p>
            )
        }
    </div>
);

const mapStateToProps = (state, props) => {
    return {
        expenseCount: props.expenses.length, // number of expenses after filter.
        expensesTotal: numeral(selectExpensesTotal(props.expenses) / 100).format('$0,0.00') 
    };
};

// UDEMY VERSION:
// 1. Summary Component inside Dashboard without any props from parent (only state from a store)
// 2. He selectExpenses here again (doubled with ExpenseList)
// 3. He destruct props and use destructed variables in component (instead props.expensesTotal)
import selectExpenses from '../selectors/expenses';

export const ExpensesSummaryUdemy = ({ expenseCount, expensesTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses' ;
    const formatedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

    return (
        <div>
            <h2>Viewing {expenseCount} {expenseWord} totalling {formatedExpensesTotal} </h2>
        </div>
    );
};

const mapStateToPropsUdemy = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

//DEFAULT UNNAMED EXPORT:
export default connect(mapStateToPropsUdemy)(ExpensesSummaryUdemy);