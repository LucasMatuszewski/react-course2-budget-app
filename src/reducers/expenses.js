/***********
 * REDUCERS
 */

// Expanses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => { //default state is empty array
    switch (action.type) {
        case 'ADD_EXPENSE':
            // state.push(action.expense); // We can't use PUSH = it changes array. Reducer wold not be a PURE Function
            // state.concat(action.expense); // CONCAT() don't change array, it returns new one.
            // BUT WE CAN USE ES6 SPREAD OPERATOR:
            return [ //with spread operator we have a control where to add new item (it could be on beginning)
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            // console.log(action.id);
            // console.log(state);
            // function removeId(expense) {
            //     console.log(expense.id);
            //     return expense.id != action.id
            // }
            // return state.filter((expense) => expense.id != action.id );
            return state.filter(({ id }) => id != action.id); // destructuring id from expense object
        // filter doesn't change array, returns new one
        case 'EDIT_EXPENSE':
            /* //my idea to do it. Works, but .map() is better :)
            console.log('state:', state.filter(({ id }) => id == action.id)[0]);
            return {
                ...state.filter(({id}) => id == action.id )[0],
                amount: action.updates
            }; */
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense, //spread previous values
                        ...action.updates //spread with new values, to edit / overwrite old values
                    }
                } else {
                    return expense;
                }
            });
        // INSTRUCTOR DIDN'T WANT US TO ADD THIS IN EXPENSES REDUCER... ONLY TO FILTERS REDUCER
        /* case 'SET_TEXT_FILTER':
            return state.filter(({ id, description, note, amount }) => {
                return id == action.text || description == action.text || note == action.text || amount == action.text;
                // return description == action.text;
            }); */
        default:
            return state;
    }
};

// export default expensesReducer; // we don't have to name this reducer if we export it by default