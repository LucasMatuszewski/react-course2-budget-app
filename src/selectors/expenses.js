import moment from 'moment';
/******
 * SELECTOR
 * GET FILTERED/Visable EXPENSES & Sort them
 */
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        /* OLD VERSION WITHOUT MOMENT():
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof startDate !== 'number' || expense.createdAt <= endDate; */
        // QUERY with MOMENT() https://momentjs.com/docs/#/query/is-same-or-before/
        // IMPORTANT!!! There is some error/bug because of Redux DevTools on data changes:
        // startDate.isSameOrBefore is not a function
        // After deactivation of Redux DevTools it works fine.
        // similar isue: https://github.com/zalmoxisus/redux-devtools-extension/issues/315
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true; // startDate is already a moment() object, we don't have to convert it
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        // const textMatch = typeof text !== 'string' || expense.description.toLowerCase().indexOf(text.toLowerCase()) !== -1 || expense.note.toLowerCase().indexOf(text.toLowerCase()) !== -1;
        //in ES7 / ES2016 we have .includes(), new method better from .indexOf() in many cases:
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase());
        // .includes() search also for NaN and undefined (indexOf don't)

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

// TIMESTAMPS (in milliseconds, 1s = 1000 ms )
// start 0 = 01.01.1970 (unix epoch)
// 1000 = 1000 milliseconds = 1s after 01.01.1970
// -60000 = 60s = 1min before 01.01.1970