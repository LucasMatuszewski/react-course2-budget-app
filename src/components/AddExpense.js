import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

// 1. THIS STATELESS COMPONENT HAVE INLINE FUNCTIONS. TO AVOID IT WE WILL REFACTOR IT TO CLASS COMPONENT:
// const AddExpense = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//             onSubmit={(expense) => {
//                 {/* Dispatch(imported action) can't be used by JEST TEST: */}
//                 {/* props.dispatch(addExpense(expense)); */}
//                 {/* WE HAVE TO MAP DISPATCH TO PROPS, and use here mapped function: */}
//                 props.onSubmit(expense);
//                 props.history.push('/'); // redirect to main page (dashboard) using browser history
//                 {/* console.log(expense); */}
//             }}
//         />
//     </div>
// );

// 2. TO AVOID INLINE FUNCTIONS WE REFACTORED STATELESS FUNCTION COMPONENT TO THIS:
export class AddExpense extends React.Component { // named export of unconnected component for JEST TESTS
    onSubmit = (expense) => {
        this.props.addExpense(expense); // action generator
        this.props.history.push('/'); // redirect to main page (dashboard) using browser history
    }
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
};

// Normal Component without Redux gets props only from parent Component and send props to children.
// With mapStateToProps we can connect current state of a component to Redux Store
// With mapDispatchToProps we can connect dispatching ACTIONS to Redux Store ???
// Now, our component dispatch actions and use state only by Redux Store.
// It gives us separation of concerns: Components are only to Display. Fetching data / actions are for Redux.
// We should also split components to:
// * Container (fetch data from Redux Store and send it to Presentational Child Component)
// * Presentational (display data from Container Parent Component)
// https://redux.js.org/basics/usage-with-react#presentational-and-container-components


// We don't need mapStateToProps here - we don't take data from props (e.g. existing expense)

// BUT for Jest Test we need mapDispatchToProps because we want to test:
// props.dispatch(addExpense(expense));
// But Jest don't have access to dispatch. With mapping, we can use props.onSubmit(expense) in Component.
// So lets map dispatch to our props:
/* const mapDispatchToProps = (dispatch) => {
    return {
        onSubmit: (expense) => dispatch(addExpense(expense))
    }; // NOW WE CAN USE: props.onSubmit(expense);
}; */

//SHORTER VERSION:
const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
}); // NOW WE CAN USE: props.onSubmit(expense) and TEST it with JEST SPY.

// MORE ABOUT CONNECT AND MAPPING:
// https://github.com/reduxjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options
// https://stackoverflow.com/questions/39419237/what-is-mapdispatchtoprops

export default connect(undefined, mapDispatchToProps)(AddExpense);
// export default connect(STATE = UNDEFINED, DISPATCH)(COMPONENT TO CONNECT WITH REDUX STORE);