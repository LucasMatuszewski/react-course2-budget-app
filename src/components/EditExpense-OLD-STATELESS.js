// OLD VERSION WITH STATELESS FUNCTION COMPONENT, without mapDispatchToProps, and named exporting to JEST test
// REFACTOR MADE THE SAME WAY LIKE FOR AddExpense.js Component.

import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
// import { removeExpense } from '../actions/expenses'; // we can import two in one line

// We have access to Props from <Route /> in AppRouter.js (like with normal React Component)

/* 
// We can use this short version, if we don't use any JS in Component (only wants to render JSX)
const EditExpense = (props) => (
    <div>
        Edit Expense
    </div>
); 

// If we want to use some JS, we have to use normal version with return(): */
const EditExpense = (props) => {
    /* // MY VERSION - filter() returns Array after filter. find() returns first ELEMENT that match!
    // https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Obiekty/Array/find
    // https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Obiekty/Array/filter
    const expense = props.expenses.filter(({ id }) => id == props.match.params.id)[0];
    console.log('PROPS:', props);
    console.log('Expense:', expense); */

    //console.log('PROPS from EditExpense.js:', props);
    return(
        <div>
            <h1>Edit Expense: {props.expense.description}</h1>
            <p>You are editing expense id: {props.expense.id}</p>
            <ExpenseForm
                expense={props.expense}
                onSubmit={ (expense) => {
                    {/* MY VERSION:
                    // props.dispatch(editExpense(props.match.params.id, expense)); */}
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push('/');
                }}
            />
            {/* REFACTOR: moved "remove" button from ExpenseListItem.js */}
            <button onClick={() => {
                props.dispatch(removeExpense({id: props.expense.id})); {/* ES6 Object definition shorthand for "id: id" */}
                props.history.push('/');
            }}>
                Remove
            </button>
        </div>
    );
};
/* 
In props we have access to info from React-Router:
* Browser History of URLs (with goBack, goForward), 
* Current location, with pathname
    * Search value (if we use ?query=xxx&sort=xxx in URL or ?anything=xxx)
    * Hash value from URL (if we use edukey.us#header)
* Path : URL Match - isExact: true/false
*/

// const mapStateToProps = (state) => { // MY VERSION
const mapStateToProps = (state, props) => { // UDEMY VERSION
    return {
        // expenses: state.expenses // MY VERSION
        /* expense: state.expenses.find((expense) => { // UDEMY VERSION. I used Filter inside this component
            return expense.id === props.match.params.id;
        }) */
        expense: state.expenses.find((expense) => expense.id === props.match.params.id) //SHORTER VERSION from UDEMY
    };
};

export default connect(mapStateToProps)(EditExpense);