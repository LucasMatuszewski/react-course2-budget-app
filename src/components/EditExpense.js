// NEW VERSION WITH STATEFULL CLASS COMPONENT, mapDispatchToProps and named exporting to JEST test
// REFACTOR MADE THE SAME WAY LIKE FOR AddExpense.js Component.
// OLD VERSION WITH MORE COMMENTS INSIDE: EditExpense-OLD-STATELESS.js

import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
// import { removeExpense } from '../actions/expenses'; // we can import two in one line



export class EditExpense extends React.Component {
    onSubmit = (expense) => {
        // this.props.editExpense(expense); // MY VERSION (works fine...)
        this.props.editExpense(this.props.expense.id, expense); // UDEMY VERSION
        this.props.history.push('/');
    }
    onRemove = () => {
        // this.props.removeExpense(); // MY VERSION (works fine...)
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }
    render() {
        return(
            <div>
                <h1>Edit Expense: {this.props.expense.description}</h1>
                <p>You are editing expense id: {this.props.expense.id}</p>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={ this.onSubmit }
                />
                <button onClick={ this.onRemove }>
                    Remove
                </button>
            </div>
        );
    }
};
/* 
In props we have access to info from React-Router:
* Browser History of URLs (with goBack, goForward), 
* Current location, with pathname
    * Search value (if we use ?query=xxx&sort=xxx in URL or ?anything=xxx)
    * Hash value from URL (if we use edukey.us#header)
* Path : URL Match - isExact: true/false
*/

// const mapStateToProps = (state) => { // MY VERSION
const mapStateToProps = (state, props) => { // UDEMY VERSION
    return {
        // expenses: state.expenses // MY VERSION
        /* expense: state.expenses.find((expense) => { // UDEMY VERSION. I used Filter inside this component
            return expense.id === props.match.params.id;
        }) */
        expense: state.expenses.find((expense) => expense.id === props.match.params.id) //SHORTER VERSION from UDEMY
    };
};
/* 
// MY VERSION (worked fine, but why???)
const mapDispatchToProps = (dispatch) => ({
    editExpense: (expense) => dispatch(editExpense(props.expense.id, expense)),
    removeExpense: () => dispatch(removeExpense({id: props.expense.id}))
}); // NOW WE CAN USE: props.onSubmit(expense) and TEST it with JEST SPY.
*/

//UDEMY VERSION:
const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
}); // NOW WE CAN USE: props.onSubmit(expense) and TEST it with JEST SPY.


export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);