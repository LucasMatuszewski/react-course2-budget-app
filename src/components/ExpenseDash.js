import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDash = () => (
    <div>
        <h2>Expenses Dashboard</h2>

        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList /> {/* We don't need to pass any data/props here. connect() will pass them directly from a store / state tree */}
        <br />
        <hr />
        <a href="/help">Server Side Rendering HTML Link</a><br />
        (html "a" tag will reload a browser, we can use it to link to external page)
    </div>
);

export default ExpenseDash;