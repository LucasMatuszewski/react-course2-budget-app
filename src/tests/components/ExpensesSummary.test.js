import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary, ExpensesSummaryUdemy } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render Expenses Summary with 3 expenses totalling $1,500.00', () => {
    const wrapper = shallow(<ExpensesSummary expensesTotal='$1,500.00' expenseCount='3' />);
    expect(wrapper).toMatchSnapshot();
});

test('should render Expenses Summary with 1 expense totalling $100.00', () => {
    const wrapper = shallow(<ExpensesSummary expensesTotal='$100.00' expenseCount='1'  />);
    expect(wrapper).toMatchSnapshot();
});

// UDEMY VERSION:
test('should render Expenses Summary UDEMY with 3 expenses totalling $1,500.00', () => {
    const wrapper = shallow(<ExpensesSummaryUdemy expensesTotal='150000' expenseCount='3' />);
    expect(wrapper).toMatchSnapshot();
});

test('should render Expenses Summary UDEMY with 1 expense totalling $100.00', () => {
    const wrapper = shallow(<ExpensesSummaryUdemy expensesTotal='10000' expenseCount='1' />);
    expect(wrapper).toMatchSnapshot();
});