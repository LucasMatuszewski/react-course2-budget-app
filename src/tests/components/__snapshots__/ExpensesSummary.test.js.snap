// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expenses Summary UDEMY with 1 expense totalling $100.00 1`] = `
<div>
  <h2>
    Viewing 
    1
     
    expenses
     totalling 
    $100.00
     
  </h2>
</div>
`;

exports[`should render Expenses Summary UDEMY with 3 expenses totalling $1,500.00 1`] = `
<div>
  <h2>
    Viewing 
    3
     
    expenses
     totalling 
    $1,500.00
     
  </h2>
</div>
`;

exports[`should render Expenses Summary with 1 expense totalling $100.00 1`] = `
<div>
  <p>
    Total: 
    $100.00
    , from 
    1
     expense
    s
  </p>
</div>
`;

exports[`should render Expenses Summary with 3 expenses totalling $1,500.00 1`] = `
<div>
  <p>
    Total: 
    $1,500.00
    , from 
    3
     expense
    s
  </p>
</div>
`;
